### PR TITLE
Add kube-controller manager API round trip test

### PR DESCRIPTION
**Pull request by stewart-yu**
**What this PR does / why we need it**:
Add kube-controller manager round trip test

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
xref #68244 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by stewart-yu** [comment]
does we need add SIG information into the OWNER file? @nikhita

**Comment by nikhita** [comment]
> does we need add SIG information into the OWNER file? @nikhita

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/apis/config/types.go has types related to multiple sigs. Example: `CloudProviderConfiguration` is owned by SIG Cloud Provider. Since this area of the code is cross-cutting, I don't think it the OWNERS file should contain any sig labels.

**Review by hzxuzhonghu — APPROVED** [review]

File: staging/src/k8s.io/kube-controller-manager/config/v1alpha1/types.go [hunk 1]
```diff
@@ -22,6 +22,27 @@ import (
 	apiserverconfigv1alpha1 "k8s.io/apiserver/pkg/apis/config/v1alpha1"
 )
 
+const (
+	// DefaultAddress defines default address that kube-controller-manager's http service serve on (set to 0.0.0.0 for all interfaces).
+	DefaultAddress = "0.0.0.0"
```
[the remark this answers is not in the document]
Comment (stewart-yu) [reply staging/src/k8s.io/kube-controller-manager/config/v1alpha1/types.go]: removed

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			obj.Generic.MinResyncPeriod = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Hour}
+			obj.Generic.ClientConnection.ContentType = "application/vnd.kubernetes.protobuf"
+			obj.Generic.ClientConnection.QPS = c.Float32()
+			obj.Generic.ClientConnection.Burst = c.Int31()
+			obj.Generic.ControllerStartInterval = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			obj.Generic.LeaderElection.LeaseDuration = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			obj.Generic.LeaderElection.RenewDeadline = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			obj.Generic.LeaderElection.RetryPeriod = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			obj.Generic.LeaderElection.ResourceLock = "endpoints"
+			obj.Generic.LeaderElection.LeaderElect = c.RandBool()
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: aren't all these non-custom-format random values set by `c.FuzzNoCustom` above already?

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			obj.KubeCloudShared.NodeMonitorPeriod = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			obj.KubeCloudShared.ClusterName = "kubernetes"
+			obj.KubeCloudShared.ConfigureCloudRoutes = c.RandBool()
+			obj.KubeCloudShared.RouteReconciliationPeriod = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			// add ServiceControllerConfiguration values
+			obj.ServiceController.ConcurrentServiceSyncs = c.Int31()
+			// add AttachDetachControllerConfiguration values
+			obj.AttachDetachController.ReconcilerSyncLoopPeriod = metav1.Duration{Duration: time.Duration(c.Int63()) * time.Second}
+			// add CSRSigningControllerConfiguration values
+			obj.CSRSigningController.ClusterSigningCertFile = "/etc/kubernetes/ca/ca.pem"
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: why is this is not a random string?

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 2 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: But in that case, https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/apis/config/v1alpha1/defaults.go#L51, we shoud set value like that `obj.TTLAfterFinishedController.ConcurrentTTLSyncs = c.Int31()`

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 3 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: If set random string, will failed. 
```
roundtrip.go:228: 	round tripping to kubecontrollermanager.config.k8s.io/v1alpha1, Kind=KubeControllerManagerConfiguration v1alpha1.KubeControllerManagerConfiguration
	roundtrip.go:338: KubeControllerManagerConfiguration: diff: 
		object.PersistentVolumeBinderController.VolumeConfiguration.FlexVolumePluginDir:
		  a: ""
		  b: "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/"
```

Comment (sttts) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: non-empty random strings should roundtrip.

Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: using `/+random string` now

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+			c.FuzzNoCustom(obj)
+			obj.Generic.Address = fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))
+			obj.Generic.ClientConnection.ContentType = fmt.Sprintf("%s/%s.%s.%s", c.RandString(), c.RandString(), c.RandString(), c.RandString())
+			obj.Generic.LeaderElection.ResourceLock = "endpoints"
+			obj.Generic.Controllers = []string{"*"}
+			obj.KubeCloudShared.ClusterName = "kubernetes"
+			obj.CSRSigningController.ClusterSigningCertFile = fmt.Sprintf("/%s", c.RandString())
+			obj.CSRSigningController.ClusterSigningKeyFile = fmt.Sprintf("/%s", c.RandString())
+			obj.PersistentVolumeBinderController.VolumeConfiguration.FlexVolumePluginDir = fmt.Sprintf("/%s", c.RandString())
+			obj.TTLAfterFinishedController.ConcurrentTTLSyncs = c.Int31()
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: this is the size I expected for this func :)

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// Funcs returns the fuzzer functions for the kube-controller manager apis.
+func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+	return []interface{}{
+		func(obj *kubectrlmgrconfig.KubeControllerManagerConfiguration, c fuzz.Continue) {
+			c.FuzzNoCustom(obj)
+			obj.Generic.Address = fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))
+			obj.Generic.ClientConnection.ContentType = fmt.Sprintf("%s/%s.%s.%s", c.RandString(), c.RandString(), c.RandString(), c.RandString())
+			obj.Generic.LeaderElection.ResourceLock = "endpoints"
+			obj.Generic.Controllers = []string{"*"}
+			obj.KubeCloudShared.ClusterName = "kubernetes"
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: this cannot be random?

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+// Funcs returns the fuzzer functions for the kube-controller manager apis.
+func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+	return []interface{}{
+		func(obj *kubectrlmgrconfig.KubeControllerManagerConfiguration, c fuzz.Continue) {
+			c.FuzzNoCustom(obj)
+			obj.Generic.Address = fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))
+			obj.Generic.ClientConnection.ContentType = fmt.Sprintf("%s/%s.%s.%s", c.RandString(), c.RandString(), c.RandString(), c.RandString())
+			obj.Generic.LeaderElection.ResourceLock = "endpoints"
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: this cannot be random?

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+			obj.Generic.Controllers = []string{"*"}
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: this cannot be random?

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 6 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: yes, error like that:
```
roundtrip.go:338: KubeControllerManagerConfiguration: diff: 
		object.Generic.LeaderElection.ResourceLock:
		  a: ""
		  b: "endpoints"
```

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 5 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: yes, error like that `obj.Generic.LeaderElection.ResourceLock`, as below

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 7 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: fix

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 6 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: I'm guess that: `object.Generic.LeaderElection.ResourceLock` in `default.go` must not-empty, so we should also set not-empty random value for `object.Generic.LeaderElection.ResourceLock` in `Funcs`. The same `obj.KubeCloudShared.ClusterName`

**Comment by stewart-yu** [comment]
@sttts  can we get it merged?:wink:

File: pkg/controller/apis/config/fuzzer/fuzzer.go (hunk 5 above)
Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: Fixed, add judgement like that:
```
if obj.KubeCloudShared.ClusterName == "" {
    obj.KubeCloudShared.ClusterName = "kubernetes"
}
```

File: pkg/controller/apis/config/fuzzer/fuzzer.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+	return []interface{}{
+		func(obj *kubectrlmgrconfig.KubeControllerManagerConfiguration, c fuzz.Continue) {
+			c.FuzzNoCustom(obj)
+			obj.Generic.Address = fmt.Sprintf("%d.%d.%d.%d", c.Intn(256), c.Intn(256), c.Intn(256), c.Intn(256))
+			obj.Generic.ClientConnection.ContentType = fmt.Sprintf("%s/%s.%s.%s", c.RandString(), c.RandString(), c.RandString(), c.RandString())
+			if obj.Generic.LeaderElection.ResourceLock == "" {
+				obj.Generic.LeaderElection.ResourceLock = "endpoints"
+			}
+			obj.Generic.Controllers = []string{fmt.Sprintf("%s", c.RandString())}
```
Comment (sttts) [inline pkg/controller/apis/config/fuzzer/fuzzer.go]: isn't this defaulted to some random string list?

Comment (stewart-yu) [reply pkg/controller/apis/config/fuzzer/fuzzer.go]: Yes. some format like that: ["*"], ["bar", "foo"] and so on

**Pull request merged**